base/task.jl

```diff
@@ -365,6 +365,86 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+
+    for t in waiting_tasks
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+    end
+
+    if all && !failfast
+        # Force everything to finish synchronously for the case of waitall
+        # with failfast=false
+        for t in tasks
+            _wait(t)
+        end
+        return tasks, Task[]
+    end
```

Comment:
@vtjnash I've added this loop as the answer to your comment https://github.com/JuliaLang/julia/pull/53341#discussion_r1509354697.
I think this is better than inserting a fully synchronous path in the loop below.